prow/pubsub/subscriber/cache.go

```diff
@@ -0,0 +1,311 @@
+	if err != nil {
+		return nil, false, false, err
+	}
+	key, err := keyConstructor()
+	if err != nil {
+		return nil, false, false, err
+	}
+	evicted = p.presubmits.Add(key.String(), presubmits)
+
+	return presubmits, false, evicted, nil
+}
+
+// GetPostsubmitsFromCache is virtually identical to GetPresubmitsFromCache. The
+// only real difference is in the keyConstructor (postsubmits don't consider
+// headSHAGetters).
+func (p *ProwYAMLCache) GetPostsubmitsFromCache(
+	vg func(git.ClientFactory, string, config.RefGetter, ...config.RefGetter) ([]config.Postsubmit, error),
+	gc git.ClientFactory,
+	identifier string,
+	baseSHAGetter config.RefGetter) ([]config.Postsubmit, bool, bool, error) {
+
+	keyConstructor := func() (fmt.Stringer, error) {
+		return MakeCacheKey(identifier, baseSHAGetter)
+	}
+
+	valConstructor := func() (interface{}, error) {
+		return vg(gc, identifier, baseSHAGetter)
+	}
+
+	val, cacheHit, evicted, err := GetFromCache(p.postsubmits, keyConstructor, valConstructor)
+	if err != nil {
+		return nil, cacheHit, evicted, err
+	}
+
+	if postsubmits, ok := val.([]config.Postsubmit); ok {
+		return postsubmits, cacheHit, evicted, err
+	}
+
+	postsubmits, err := vg(gc, identifier, baseSHAGetter)
+	if err != nil {
+		return nil, false, false, err
+	}
+	key, err := keyConstructor()
+	if err != nil {
+		return nil, false, false, err
+	}
+	evicted = p.presubmits.Add(key.String(), postsubmits)
+
+	return postsubmits, false, evicted, nil
+}
+
+// GetFromCache tries to use a cache if it is available to get a Value. It is
+// assumed that Value is expensive to construct from scratch, which is the
+// reason why we try to use the cache in the first place. If we do end up
+// constructing a Value from scratch, then we store it into the cache with a
+// corresponding Key, so that we can look up the Value with just the Key in the
+// future.
+func GetFromCache(
+	cache *lru.Cache,
+	keyConstructor keyConstructor,
+	valConstructor valConstructor) (interface{}, bool, bool, error) {
+
+	// If the cache is unreachable, then fall back to cache-less behavior
+	// (construct the value from scratch).
+	if cache == nil {
+		valConstructed, err := valConstructor()
+		if err != nil {
+			return nil, false, false, err
+		}
+
+		return valConstructed, false, false, nil
+	}
+
+	// Construct cache key. We use this key to find the value (if it was already
+	// stored in the cache by a previous call to GetFromCache).
+	key, err := keyConstructor()
+	if err != nil {
+		return nil, false, false, err
+	}
+
+	// Cache lookup.
+	valFound, ok := cache.Get(key.String())
+
+	// Cache hit.
+	if ok {
+		return valFound, true, false, nil
+	}
+
+	// Cache miss. We have no choice but to construct the value (this call is
+	// expensive!) and add it to the cache.
+	valConstructed, err := valConstructor()
+	if err != nil {
+		return nil, false, false, err
+	}
+
+	// Add our constructed value to the cache.
+	evicted := cache.Add(key.String(), valConstructed)
```

Comment:
I think the cache miss behavior is wrong for our use case. In particular, this cache will frequently encounter bursts of nearly simultaneous requests for new keys that we haven't cached yet. When that happens we want to calculate the value once and return that to all consumers, but the current behavior for a cache miss is to just calculate the value and then add it to the cache, even if multiple other threads are doing the same thing concurrently. Calculating the value once for concurrent cache misses is called request coalescing and `ghproxy` implements it if you are looking for an example.
I think this is something that needs to be supported at the cache locking layer so hopefully the LRU library supports this. If not, we could either add a redundant lock layer that we control, or we could make a very simple approximation of an LRU as we discussed before and use that instead.